pixeltable/exprs/expr.py

```diff
@@ -483,3 +496,87 @@ def _make_arithmetic_expr(self, op: ArithmeticOperator, other: object) -> 'pixel
             return ArithmeticExpr(op, self, Literal(other))  # type: ignore[arg-type]
         raise TypeError(f'Other must be Expr or literal: {type(other)}')
 
+    def _make_applicator_function(self, fn: Callable, col_type: Optional[ts.ColumnType]) -> 'pixeltable.func.Function':
+
+        if col_type is not None:
+            # col_type is specified explicitly
+            fn_type = col_type
+        elif fn in _known_applicator_types:
+            # For convenience, various built-ins and other Python functions that don't
+            # have type hints are hardcoded
+            fn_type = _known_applicator_types[fn]
+        elif 'return' in typing.get_type_hints(fn):
+            # Attempt to infer the column type from the return type of the callable;
+            # this will set fn_type to None if it cannot be inferred
+            return_type = typing.get_type_hints(fn)['return']
+            fn_type = ts.ColumnType.from_python_type(return_type)
+        else:
+            # No type hint
+            fn_type = None
+
+        if fn_type is None:
+            raise excs.Error(
+                f'Column type of `{fn.__name__}` cannot be inferred. Use `.apply({fn.__name__}, col_type=...)` to specify.'
+            )
+
+        # TODO(aaron-siegel) Currently we assume that `fn` has exactly one required parameter
+        # and all optional parameters take their default values. Should we provide a more
+        # flexible API? For example, by defining
+        # expr.apply(fn, my_kw=my_arg)
+        # to mean: transform each x by calling
+        # fn(x, my_kw=my_arg)
+        # In the current implementation, a lambda is needed in order to specify this pattern:
+        # expr.apply(lambda x: fn(x, my_kw=my_arg))
+
+        try:
+            # If `fn` is not a builtin, we can do some basic validation to ensure it's
+            # compatible with `apply`.
+            params = inspect.signature(fn).parameters
+            params_iter = iter(params.values())
+            first_param = next(params_iter) if len(params) >= 1 else None
+            second_param = next(params_iter) if len(params) >= 2 else None
+            # Check that fn has at least one positional parameter
+            if len(params) == 0 or first_param.kind in (inspect.Parameter.KEYWORD_ONLY, inspect.Parameter.VAR_KEYWORD):
+                raise excs.Error(
+                    f'Function `{fn.__name__}` has no positional parameters.'
+                )
+            # Check that fn has at most one required parameter, i.e., its second parameter
```

Comment:
So we're okay with 
def f(x=1, y=2):
...
t.col.apply(f)?

Feels a bit weird. I'm wondering if we should require exactly one required parameter.

Reply:
Yeah I considered this case and intentionally decided to support it. It's the most logical behavior to me - if I were a user, I'd think it was weird for apply() to reject a function just because its first parameter has a default. I don't feel super strongly about it though.